### PR TITLE
Image: Reimplement lightbox trigger as a minimal button in corner of image

### LINKED ISSUE #55097: Image block lightbox: do not overlay the image with a transparent button
### Description

When an image block is set to open the image in the lightbox, a transparent button entirely overlays the image.

This button is necessary to open the modal dialog. However, it entirely overlays the image. With the current implementation I see a couple problems that I'd consider pretty serious.

1
**The browsers contextual menu that opens on right-click is different depending on the browser in use.**

This surprised me. I would have expected to always get the same contextual menu. I have no idea whether this is a browser bug or a 'feature'. Regardless:
- Chrome and Opera open the _image_ contextual menu. This looks 'visually correct' but it's wrong, as the element that actually receives the right-click is the transparent button.
- Firefox and Safari open the _button_ contextual menu. This looks 'visually wrong' but it is actually the correct behavior.

Regardless which browser is right and which one is wrong, the current implementation triggers a behavior that breaks am expected browsers feature. Also, the user experience is pretty confusing in at least two of the major browsers:
- As an user, I see a normal image.
- When right clicking, I would expect to be able to save the image, open it in a new tab, copy the image, etc.
- Instead, the browser contextual menu doesn't show me any of those commands.

2
**There is no visual indication at all that the image can be enlarged in a modal dialog**

Visually, the image with the lightbox looks exactly like a normal image. There's no visual indication or any other hint to communicate that the image can be clicked and enlarged in the lightbox. The only subtle visual indications are only visible when hovering or tabbing:
- Hovering: the mouse cursor changes to `zoom-in`. 
- Tabbing: the focused 'image' (actually it's the button) shows a focus outline.

Seems to me both cases are insufficient indications.
- On mobile there is no hovering or tabbing. I wonder how users are supposed to understand the image can be enlarged.
- On mobile, users will likely think the image is a normal image. Potentially, they could trigger an unwanted click when scrolling the page and the lightbox would open unexpectedly.
- Also on desktop understanding the image is clickable is not trivial.
- Some users may use devices or assistive technologies that don't use hover or tabbing.
- The mouse cursor change is an unreliable indication as some users may have set a custom cursor at the operating system level.

### Step-by-step reproduction instructions

- Add an image block and set it to open in the lightbox.
- Publish and view the front end.
- Test with macOS Chrome and Opera:
  - Right-click the 'image' (reminder that you are actually right-clicking the transparent button)
  - Observe the contextual menu that opens is the one related to the image.
- Test with macOS Firefox and Safari:
  - Right-click the 'image' (reminder that you are actually right-clicking the transparent button)
  - Observe the contextual menu that opens is the one related to the button.

### Screenshots, screen recording, code snippet

In the screenshots below, I altered the transparent button to better illustrate the issue:
- I applied a semi-transparent red background to the button to make it visible while keeping the image visible.
- I shifted the button a little to the right and bottom, to be able to right-click the image behind the button.

macOS Chrome and Opera: right-click on the image shows the image contextual menu, as expected:

chrome opera 01

macOS Chrome and Opera: right-click on the button shows the image contextual menu, this is not expected:

chrome opera 02

macOS Firefox and Safari: right-click on the image shows the image contextual menu, as expected:

firefox safari 01

macOS Firefox and Safari: right-click on the button shows the button contextual menu, as expected:

firefox safari 02

## Proposal

I'd think the simplest and most effective solution would be to just not entirely overlay the transparent button over the image. A visible button with an appropriate icon and labelling (preferably an icon + visible text for a11y reasons) would solve both issues:
- right clicking on the image would be consistent in all browsers
- there would be a clear, visible, indication that the image can be enlarged

I think a button in the top right is a pretty common pattern I've seen being used around on the web. Basically, a button placed in the red spot highlighted in the screenshot below:

Screenshot 2023-10-03 at 16 28 13

Ideally, the buton should have visible text:

`Enlarge {icon here}`

In all cases, the button size should be at least 44 by 44 pixels.

### Environment info

_No response_

### Please confirm that you have searched existing issues in the repo.

Yes

### Please confirm that you have tested with all plugins deactivated except Gutenberg.

Yes

### PR DESCRIPTION
**Pull request by artemiomorales**
## What?

This PR changes the lightbox trigger so that it exists in the top right corner of the image rather than overlaying it.

## Why?

Addresses https://github.com/WordPress/gutenberg/issues/55097
Having the lightbox overlay the entire image interrupts key browser functionality and is prone to failing with certain themes and layouts.

## How?

The PR changes the styles of the button and removes the JavaScript handling that was calculating the button position dynamically.

## Testing Instructions

1. Add an image to a post
2. Enable 'Expand on Click' in the image block settings
3. Publish and view the post
4. Hover over the image with a mouse and see that the lightbox trigger becomes visible in the top right corner. Notice that hovering over the button will cause the cursor to become a pointer. Try clicking on the button and ensure the lightbox opens.
5. Notice also that hovering over the image causes the cursor to become a magnifying glass. Click on the image and ensure the lightbox opens.

Additionally:
- Try adding a `title` to the image. Hover over the image and ensure the `title` appears in a tooltip.
- Try right clicking on an image and ensuring the image contextual menu appears as expected in Chrome, Safari, Firefox, Opera.

### Testing Instructions for Keyboard

1. While viewing the post, tab to the image and make sure the lightbox trigger appears on focus. 
2. Press `Enter` or `Spacebar` to ensure the lightbox opens as expected.

## Screenshot
Screenshot 2023-10-10 at 10 36 50 AM

### CODE REVIEW

**Comment by artemiomorales** [comment]
I'm noting here that if we proceed with this, as part of it, we'll also be able to add an additional style declaration that would resolve the bug wherein theme styles conflict with the lightbox button in https://github.com/WordPress/gutenberg/issues/54682

**Comment by michalczaplinski** [comment]
I think that showing a button on hover only is a decent compromise, nice work @artemiomorales 👍 

I see 2 issues here though: 

1. When using some custom aspect ratios, the button can appear outside the image: 

   https://github.com/WordPress/gutenberg/assets/5417266/08153050-d334-4a81-83a5-531db4a14e81

2. Images with rounded borders look a bit weird having the button in the corner which is actually outside of the image. 

   Screenshot 2023-10-11 at 16 32 53
   
   Edge cases like this make me question a bit whether having this button is worth the trouble, but I don't want to reopen the discussion that was already had in https://github.com/WordPress/gutenberg/issues/55097. If we can fix 1., I'm fine with shipping the button in the corner of the image.

**Comment by artemiomorales** [comment]
@michalczaplinski Great, thanks! I've just pushed a commit covering edge cases when using `contain`. Next step is to address scenarios wherein the image has custom width or height properties.

File: packages/block-library/src/image/style.scss
Comment (jameskoster) [inline packages/block-library/src/image/style.scss]: Unsure about conventions, but should this be `left` for `rtl` languages?

File: packages/block-library/src/image/style.scss
Comment (jameskoster) [inline packages/block-library/src/image/style.scss]: Not sure we should use a UI color for this button. It can probably just be black to begin with. In the future we might consider making the appearance consistent with other Buttons as defined by the active theme.

File: packages/block-library/src/image/style.scss
Comment (jameskoster) [inline packages/block-library/src/image/style.scss]: The icon footprint should be 24px, the button can probably match.

File: packages/block-library/src/image/view.js
Comment (SantosGuillamot) [inline packages/block-library/src/image/view.js]: Can't we use [the next-sibling combinator](https://developer.mozilla.org/en-US/docs/Web/CSS/Next-sibling_combinator) with something like this:

```CSS
img:hover + button {
    opacity: 1;
}
```

Using JS for this seems a bit complex for this use case to me.

File: packages/block-library/src/image/view.js
Comment (SantosGuillamot) [inline packages/block-library/src/image/view.js]: Can't we store the `` element ref in the context and use it [here](https://github.com/WordPress/gutenberg/blob/cea71f1e57712d370e503e5232a806e6fda562ed/packages/block-library/src/image/view.js#L109)? Something like:

```js
setStyles( context, context.core.image.imageRef );
```

It feels like we are creating different actions serving the same purpose.

**Comment by SantosGuillamot** [comment]
I've been taking a look at this, and I have some comments so far:

* While hovering the button cursor is `pointer` and in the image is `zoom-in`. Is this expected?

Screenshot 2023-10-16 at 17 43 58

Screenshot 2023-10-16 at 17 43 40

* In the images with the "Contain" setting activated, the cursor shows as zoom in. I believe this wasn't the case before and it was discussed back then.

Screenshot 2023-10-16 at 17 45 37

* In small images, the button is a bit weird.

Screenshot 2023-10-16 at 17 46 59

Screenshot 2023-10-16 at 17 47 17

**Comment by artemiomorales** [comment]
@SantosGuillamot 

> While hovering the button cursor is pointer and in the image is zoom-in. Is this expected?

Yes, this is expected.

> In the images with the "Contain" setting activated, the cursor shows as zoom in. I believe this wasn't the case before and it was discussed back then.

You mean that the cursor shows as a magnifying glass even when not over the image contents, right? I'm not sure there's a good way around this. Before, we could dynamically set the button dimensions to make sure the actionable area was only the size of the contained image, but if we're not doing that, then the `img` element itself is actually just bigger than the contained image (see below — the whole highlighted area is the `img`).

A few options come to mind:
1. Don't show the magnifying glass on hover, and don't trigger the lightbox when the image is clicked.
2. Accept this as a limitation.
3. Implement a JavaScript hack that will listen for the position of the cursor and, if it's within the bounds of the image, activate an invisible `div` whose dimensions are calculated to overlay the image content precisely and which also contains the styling for the magnifying glass.

I'm leaning towards 2, and haven't decided yet whether 3 is a good idea.

Screenshot 2023-10-16 at 2 04 00 PM

File: packages/block-library/src/image/view.js
Comment (artemiomorales) [reply packages/block-library/src/image/view.js]: You're right! I'd forgotten about that one.

File: packages/block-library/src/image/style.scss
Comment (artemiomorales) [reply packages/block-library/src/image/style.scss]: Maybe! That sounds like a good candidate for a follow up PR. I've added this point to the next version in the [lightbox tracking issue](https://github.com/WordPress/gutenberg/issues/51132).

File: packages/block-library/src/image/view.js
Comment (artemiomorales) [reply packages/block-library/src/image/view.js]: Yup, you're right! I've taken your suggestion and cleaned it up.

File: packages/block-library/src/image/style.scss
Comment (richtabor) [reply packages/block-library/src/image/style.scss]: Agree with black.

**Comment by richtabor** [comment]
Do we need 25px offset? Cutting it to `10` reduces the display issue on smaller images: 

CleanShot 2023-10-16 at 17 33 00

CleanShot 2023-10-16 at 17 33 59

**Comment by artemiomorales** [comment]
@SantosGuillamot @michalczaplinski I've just pushed a few commits that now cover all of the cases that were causing issues before. I'm going to continue testing different combinations of settings but believe this is ready for another look.

The only exception is that I still haven't resolved the issue [with certain `contain` values](https://github.com/WordPress/gutenberg/pull/55212#issuecomment-1765120709). I'm not sure there's a good solution for that, or at least not one we'll be able to get it in time for the 6.4 release candidate.

**Comment by richtabor** [comment]
Let's also use the updated fullscreen icon from #55021.

**Comment by artemiomorales** [comment]
> Do we need 25px offset? Cutting it to 10 reduces the display issue on smaller images:

@richtabor Thanks for taking a look! I just cut it to 10 pixels.

**Comment by SiobhyB** [comment]
@artemiomorales, I wanted to check whether you're still aiming to get this into today's RC1? As a reminder, the commit freeze is at 15.00 UTC today, so this would ideally need to be merged a couple of hours before that to give us time to organise the sync and commit.

**Comment by artemiomorales** [comment]
> I wanted to check whether you're still aiming to get this into today's RC1? As a reminder, the commit freeze is at 15.00 UTC today, so this would ideally need to be merged a couple of hours before that to give us time to organise the sync and commit.

@SiobhyB Yes, that's the hope! Although it needs a review and approval from someone else at this point. I've pinged a couple of folks so we'll see if they can take a look.

File: packages/block-library/src/image/view.js
Comment (SantosGuillamot) [inline packages/block-library/src/image/view.js:292]: Why don't we want to focus the close button when opening the button by clicking the image? Right now, it keeps the active element outside of the modal, which I believe is not correct, right?

Comment (artemiomorales) [reply packages/block-library/src/image/view.js:292]: That was a point brought up some folks from design a while back — they did not want a close button at all because they felt it would distract and be unnecessary.

I figured adding this so the focus ring doesn't become visible and draw attention to the close button would be a good move. Personally I would prefer to not have my attention immediately drawn to a close button when trying to get a better view of an image.

**Review by aristath — APPROVED** [review]
Let me prefix this by saying I understand this is not _yet_ perfect... I can see a couple of edge-cases above that would warrant some more polishing.
However, as a whole, this implementation is a lot better than the one we currently have.
With that in mind, and understanding that we're on the clock for RC1, I'll go ahead and pre-approve this PR. This will give us the confidence that we have something better than what was previously there, while giving us another week to polish the styles based on feedback we receive from RC1 testers 👍 

The code looks good, and this is definitely a welcomed improvement

File: packages/block-library/src/image/view.js
Comment (SantosGuillamot) [inline packages/block-library/src/image/view.js:164]: Same as with the close button: if we navigate to a lightbox with the keyboard, but I close it using a click, I lose the focus, which I assume is not correct, right?

**Review by SantosGuillamot — APPROVED** [review]
I've looked at it, and, apart from a couple of comments I left, I believe the PR does what is indicated in the description. The only issue I can see is the one described [here](https://github.com/WordPress/gutenberg/pull/55212#issuecomment-1765120709), but I believe we can keep it as a limitation for this RC and analyze it later.

Apart from that, I am a bit afraid of making big changes like these ones in a RC, but I'll let you decide on that.

**Comment by artemiomorales** [comment]
> Let's also use the updated fullscreen icon from https://github.com/WordPress/gutenberg/pull/55021.

@richtabor As we were pushing the RC1 deadline, I'll change this in a follow up PR.

File: packages/block-library/src/image/view.js
Comment (artemiomorales) [reply packages/block-library/src/image/view.js:292]: There are potential accessibility issues here, though. The context would indeed be incorrect for folks who activate the lightbox using a mouse while also using a screen reader.

Is this a use case we should handle? I'm currently searching Google for guidance around this, and will also ping @afercia @joedolson @alexstine for feedback.

Comment (SantosGuillamot) [reply packages/block-library/src/image/view.js:292]: The problem is that with that approach we are losing the focus and it is causing some issues. When opening an image by clicking on it:
1. I can't close it with the ESC key.
2. I can't close it navigating with the keyboard because I can't focus on the close button.

**Comment by SiobhyB** [comment]
I've gone ahead to cherry-pick this to the release branch in https://github.com/WordPress/gutenberg/pull/55405/commits/7bce633cfb1162a2127b561f0e1aad5e0e7fd1e0, so it'll be included in the RC. :)

**Comment by afercia** [comment]
I see this has been merged while I was reviewing it. No wirries, it can happen. My lunch break took more than expected I guess :) 

However, there are issues that need to be fixed, not only accessibility but also the fact that:
- `event.pointerType` is buggy across browsers
- differentiating behavior depending on pointing device vs. 'other' is not ok for accessibility

**Comment by afercia** [comment]
So this will need a follow up for the next RC. 

- The implementation based on `event.pointerType` is buggy, as Chrome and Opera regurn an empty string as expected when using the keyboard while Firefox and Safari return `undefined`. 
- Regardless, differentiating the behavior depending on pointing device vs. 'other' is not OK for accessibility. The modal dialog focus management should always have the same behaevior regardless of mouse, touch, keyboard, other.
- The assumption that users are _always_ either using a pointing device or other is incorrect. For example, I may want to open the modal dialog with a click and close it by pewssing the Escape key.

Depending on the flow used to open the dialog _and_ on the browser in use, this is what I observed when testing:

- When opening the dialog, focus is not transferred to the dialog. As such:
  - Focus stays on the page behind the dialog. Tabbing will navigate the page content behing the modal dialog.
  - The dialgo is not announced by screen readers because focus is elsewhere.
  - The dialog doesn't close on Escape.
  - When closing the dialog, focus should always return to the trigger button, instead this is not the case.
  - As such, there is a full focus loss: tabbing again will restart the tab order from the document root.
- The trigger button doesn't visually expose its accessible name. This is a broadeer issue with all icon buttons ued on the front end and should be addressed holistically.
- The trigger button is only 24 by 24 pixels. While this is compliant with the current WCAG 2.2, I would like to see the button target size a little larger.
- When opening the modal dialog, Initial focus should go to the element with role=dialog instead of the Close button.
- The clickable image is problematic. I'm not sure why we want to make a HTML element have a non-native behavior. I'm not sure that would make it more usable in the first place, as it is a non expected beheavi, especially considering the following point.  It is even more problematic on mobile.
- Still, there is no visual hint that the image can be enlarged. It looks like a normal image and only when hovered or focusd shows a visual hint. As a user, looking at a page with a Lightbox image, I wouldn't have any clue that the image can be enlarged. See screnshot below: how users are supposed to understand the image can be enlarged?

Screenshot 2023-10-17 at 13 12 47

**Pull request merged**
